pinot-common/src/main/java/com/linkedin/pinot/common/partition/PartitionAssignmentGenerator.java

```diff
@@ -0,0 +1,165 @@
+
+    // get latest segment in each partition
+    for (Map.Entry<String, Map<String, String>> entry : mapFields.entrySet()) {
+      String segmentName = entry.getKey();
+      if (LLCSegmentName.isLowLevelConsumerSegmentName(segmentName)) {
+        LLCSegmentName llcSegmentName = new LLCSegmentName(segmentName);
+        String partitionId = String.valueOf(llcSegmentName.getPartitionId());
+        LLCSegmentName latestSegment = partitionIdToLatestSegment.get(partitionId);
+        if (latestSegment == null || llcSegmentName.getSequenceNumber() > latestSegment.getSequenceNumber()) {
+          partitionIdToLatestSegment.put(partitionId, llcSegmentName);
+        }
+      }
+    }
+    return partitionIdToLatestSegment;
+  }
+
+  /**
+   * Generates partition assignment for given table, using tagged hosts and num partitions
+   */
+  public PartitionAssignment generatePartitionAssignment(TableConfig tableConfig, int numPartitions) throws Exception {
+
+    // TODO: add an override which can read from znode, instead of generating on the fly
+
+    List<String> partitions = new ArrayList<>(numPartitions);
+    for (int i = 0; i < numPartitions; i++) {
+      partitions.add(String.valueOf(i));
+    }
+
+    String tableNameWithType = tableConfig.getTableName();
+    int numReplicas = tableConfig.getValidationConfig().getReplicasPerPartitionNumber();
+    RealtimeTagConfig realtimeTagConfig = getRealtimeTagConfig(tableConfig);
+
+    List<String> consumingTaggedInstances = getConsumingTaggedInstances(realtimeTagConfig);
+    if (consumingTaggedInstances.size() < numReplicas) {
+      throw new Exception(
+          "Not enough consuming instances tagged. Must be atleast equal to numReplicas:" + numReplicas);
+    }
+
+    /**
+     * TODO: We will use only uniform assignment for now
+     * This will be refactored as AssignmentStrategy interface and implementations UniformAssignment, BalancedAssignment etc
+     * {@link PartitionAssignmentGenerator} and AssignmentStrategy interface will together replace
+     * StreamPartitionAssignmentGenerator and StreamPartitionAssignmentStrategy
+     */
+    return uniformAssignment(tableNameWithType, partitions, numReplicas, consumingTaggedInstances);
+  }
+
+  /**
+   * Uniformly sprays the partitions and replicas across given list of instances
+   * Picks starting point based on table hash value. This ensures that we will always pick the same starting point,
+   * and return consistent assignment across calls
+   * @param allInstances
+   * @param partitions
+   * @param numReplicas
+   * @return
+   */
+  private PartitionAssignment uniformAssignment(String tableName, List<String> partitions, int numReplicas,
+      List<String> allInstances) {
+
+    PartitionAssignment partitionAssignment = new PartitionAssignment(tableName);
+
+    Collections.sort(allInstances);
+
+    int numInstances = allInstances.size();
+    int serverId = 0;
```

Comment:
you still want to start from a random point to accommodate multi-tenant tables now, right?